lms/djangoapps/course_blocks/transformers/library_content.py

```diff
@@ -177,3 +188,68 @@ def publish_event(event_name, result, **kwargs):
             format_block_keys,
             publish_event,
         )
+
+
+class ContentLibraryOrderTransformer(BlockStructureTransformer):
+    """
+    A transformer that manipulates the block structure by modifying the order of the
+    selected blocks within a library_content module to match the order of the selections
+    made by the ContentLibraryTransformer or the corresponding XBlock. So this transformer
+    requires the selections for the randomized content block to be already
+    made either by the ContentLibraryTransformer or the XBlock.
+
+    Staff users are *not* exempted from library content pathways/
+    """
+    WRITE_VERSION = 1
+    READ_VERSION = 1
+
+    @classmethod
+    def name(cls):
+        """
+        Unique identifier for the transformer's class;
+        same identifier used in setup.py
+        """
+        return "library_content_randomize"
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Collects any information that's necessary to execute this
+        transformer's transform method.
+        """
+        block_structure.request_xblock_fields('mode')
+        block_structure.request_xblock_fields('max_count')
+
+        ContentLibraryTransformer.set_block_analytics_summary(block_structure, modulestore())
+
+    def transform(self, usage_info, block_structure):
+        """
+        Transforms the order of the children of the randomized content block
+        to match the order of the selections made and stored in the XBlock 'selected' field.
+        """
+        for block_key in block_structure:
+            if block_key.block_type != 'library_content':
+                continue
+
+            library_children = block_structure.get_children(block_key)
+
+            if library_children:
+                state_dict = get_student_module_as_dict(usage_info.user, usage_info.course_key, block_key)
+                current_children_blocks = set(block.block_id for block in library_children)
+                current_selected_blocks = set(item[1] for item in state_dict['selected'])
+
+                # As the selections should have already been made by the ContentLibraryTransformer,
+                # the current children of the library_content block should be the same as the stored
+                # selections. If they aren't, some other transformer that ran before this transformer
+                # has modified those blocks (for example, content gating may have affected this). So do not
+                # transform the order in that case.
+                if current_children_blocks != current_selected_blocks:
+                    logger.info(
+                        u'Mismatch between the children of %s in the stored state and the actual children for user %s. '
+                        'So not transforming the order.',
```

Comment:
nit: Does "Abortin order randomization" make sense here? The new message is a bit unclear to me.
Or perhaps "Continuing without order randomization" ?

Also, I see the re-shuffling has been removed in this section. I tested a couple scenarios, and if I add new problems to the content library, a user who has already viewed the course will not see a change in problems, *unless* the problem count is changed in the randomized content block. This doesn't seem like a problem, but does it change the behavior from before?

Reply:
@kaizoku,

> nit: Does "Abortin order randomization" make sense here? The new message is a bit unclear to me.
Or perhaps "Continuing without order randomization" ?

:+1:. I will use `order transformation` as this transform re-orders the blocks to match the randomized order in selection made in the existing `ContentLibraryTransformer`.

`ContentLibraryTransformer` makes the selection,  shuffles the children blocks and then stores them the student module. `ContentLibraryOrderTransformer` retrieves the stored selections and re-orders the children blocks to match the order in the retrieved data.

> Also, I see the re-shuffling has been removed in this section. I tested a couple scenarios, and if I add new problems to the content library, a user who has already viewed the course will not see a change in problems, unless the problem count is changed in the randomized content block. This doesn't seem like a problem, but does it change the behavior from before?

In the existing implementation, the existing selection is changed only when the `count` is changed or if any of the selected blocks are no longer valid (i.e., they have been removed from the library). This PR preserves the same behaviour and does the selection and shuffling only in these 2 cases.

The goal of the changes in this PR is to shuffle the order of the selected blocks so that the order is not similar due to set ordering. Does this make sense to you?

Reply:
Yes, that makes sense @lgp171188. I think I was just seeing a bit of code removed from this section compared to the previous PR and wanted to make sure it didn't change the behavior.

Thanks for updating the logging string as well.